### PR TITLE
Rewrite SpanHelpers.IndexOfNullByte/IndexOfNullCharacter to unmanaged pointers

### PR DESCRIPTION
**Pull request by EgorBo**
Addresses https://github.com/dotnet/runtime/issues/81332

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/String.cs [hunk 1]
```diff
@@ -575,9 +575,9 @@ public StringRuneEnumerator EnumerateRunes()
             return new StringRuneEnumerator(this);
         }
 
-        internal static unsafe int wcslen(char* ptr) => SpanHelpers.IndexOfNullCharacter(ref *ptr);
+        internal static unsafe int wcslen(char* ptr) => SpanHelpers.IndexOfNullCharacter(ptr);
 
-        internal static unsafe int strlen(byte* ptr) => SpanHelpers.IndexOfNullByte(ref *ptr);
+        internal static unsafe int strlen(byte* ptr) => SpanHelpers.IndexOfNullByte(ptr);
```
Comment (am11) [inline src/libraries/System.Private.CoreLib/src/System/String.cs:580]: a nit: we can use `SpanHelpers.IndexOfNullByte` directly at call-sites and delete these wrappers to promote the usage of same API everywhere (CoreLib has total eight call-sites of `wcslen` and `strlen` combined).

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:580]: > a nit: we can use `SpanHelpers.IndexOfNullByte` directly at call-sites and delete these wrappers to promote the usage of same API everywhere (CoreLib has total 21 call-sites of `wcslen` and `strlen` combined).

It's a bit more complicated too delete it - e.g. VM relies on it for IL marshallers, mono seems too

Comment (am11) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:580]: Ah corelib.h, forgot to check that. 😁
(they are used in `ilmarshalers.cpp` via `METHOD__STRING__STRLEN` and `METHOD__STRING__WCSLEN` macros)

Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:580]: > Calls in C# code can use SpanHelpers directly.

I wonder if in this case ILLink will trim `strlen`/the other one as unused and then VM will crash, so probably better to leave as is or rename completely?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/String.cs:580]: For CoreCLR, ILLink keeps everything referenced by VM. I am not sure about Mono.

**Review by jkotas — APPROVED** [review]

**Pull request merged**